packages.js

```diff
@@ -0,0 +1,114 @@
+const debug = require('diagnostics')('warehouse:packages');
+
+const defaultOptions = {
+  cache: {}
+};
+
+class Packages {
+  /**
+   * @constructor
+   * @param {Warehouse} warehouse Reference to the Warehouse instance.
+   * @param {Object} [options] Packages options object
+   * @param {Object} [options.cache] An options object for caching package information
+   * @param {boolean} [options.cache.enabled] True if you want to use a cache of package information,
+   * False if you want to hit the service every time
+   * @param {number} [options.cache.interval] An interval in ms on which the cache should be refreshed. (Defaults to 5 minutes)
+   * @public
+   */
+  constructor(warehouse, options) {
+    this.warehouse = warehouse;
+
+    options = Object.assign({}, defaultOptions, options);
+    this.cache = new Map();
+    this.cache.interval = options.cache.interval || 5 * 60 * 1000;
+    this.cache.lastRefresh = 0;
+    this.skipCache = !options.cache.enabled;
+  }
+
+  /**
+   * Get package information for all package
+   *
+   * @param {boolean} bypassCache Should the cache be bypassed
+   * @param {Function} fn Completion callback.
+   * @returns {undefined}
+   * @private
+   */
+  _getAllPackages(bypassCache, fn) {
+    const cache = this.cache;
+
+    if (this.skipCache || bypassCache || !cache.size || (cache.lastRefresh - Date.now() > cache.interval)) {
+      return void this.warehouse.send('packages', function (err, result) {
+        if (err) return fn(err);
```

Comment:
if we can't talk to warehouse, but we have stale cache data should we just use that? (another reason to fix up `out-of-band-cache` for usage here).